obj_factory.h

```diff
@@ -0,0 +1,129 @@
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+#pragma once
+#include <vector>
+#include <unordered_map>
+#include "utils/functions.h"
+#include "utils/idx_map.h"
+
+namespace navitia{
+
+/*
+ * Factory handling all collections for one object
+ * Object must inherit from navitia::type::Header
+ */
+template<typename ObjType>
+class ObjFactory {
+public:
+    typedef typename std::vector<std::unique_ptr<ObjType>>::iterator iterator;
+    typedef typename std::vector<std::unique_ptr<ObjType>>::const_iterator const_iterator;
+
+    template<typename ...Args>
+    ObjType* emplace(const std::string& uri, Args&& ...args) {
+        check_free_uri(uri);
+        vec.push_back(std::make_unique<ObjType>(std::forward<Args>(args)...));
+        return manage_vec_back_consistency(uri);
+    }
+
+    template<typename ...Args>
+    ObjType* get_or_create(const std::string& uri, Args&& ...args) {
+        ObjType* obj = get_mut(uri);
+        if (obj == nullptr) {
+            obj = emplace(uri, args...);
+        }
+        return obj;
+    }
+
+    ObjType* insert(const std::string& uri, ObjType&& obj) {
+        check_free_uri(uri);
+        vec.push_back(std::make_unique<ObjType>(std::move(obj)));
+        return manage_vec_back_consistency(uri);
+    }
+
+    const ObjType* operator[] (const std::string& uri) const {
+        const auto obj_it = map.find(uri);
+        if (obj_it == std::end(map)) {
+            return nullptr;
+        }
+        return obj_it->second;
+    }
+
+    const ObjType* operator[] (const Idx<ObjType>& idx) const {
+        if (idx.val >= vec.size()) {
+            return nullptr;
+        }
+        return vec[idx.val];
+    }
+
+    ObjType* get_mut(const std::string& uri) {
+        auto obj_it = map.find(uri);
+        if (obj_it == std::end(map)) {
+            return nullptr;
+        }
+        return obj_it->second;
+    }
+
+    ObjType* get_mut(const Idx<ObjType>& idx) {
+        if (idx.val >= vec.size()) {
+            return nullptr;
+        }
+        return vec[idx.val];
+    }
+
+    iterator begin() {return std::begin(vec);}
+    const_iterator begin() const {return std::begin(vec);}
+    iterator end() {return std::end(vec);}
+    const_iterator end() const {return std::end(vec);}
+
+    size_t size() const noexcept {return vec.size();}
+
+    template<class Archive> void serialize(Archive & ar, const unsigned int) { ar & vec & map; }
+
+private:
+    std::vector<std::unique_ptr<ObjType>> vec;
```

Comment:
Not the expected use of IdxMap, but it's a possibility.
It's not "mandatory" here I'd say.